README.md
# DEalog Backoffice

![Elixir CI](https://github.com/DEalog/backoffice/workflows/Elixir/badge.svg)
![Build](https://github.com/DEalog/backoffice/workflows/Build/badge.svg)
![Deployment](https://github.com/DEalog/backoffice/workflows/Deployment/badge.svg)

The DEalog Backoffice is the authoring part of the DEalog platform, a platform
for official communication in cases of emergency.

For more information please see [the DEalog website](https://dealog.info).

## Development

The following section describes the development setup with all needed
prerequisites needed to get the application up and running locally.

### Requirements

- [Mutagen](https://mutagen.io) for code syncing
- [Docker](https://docker.io) for running the code
- Editor for coding

> Mutagen is mainly used to speed up file sync especially on macOS.

### Getting up and running

In order to get the stack booted just run `./dev start`. This will check if above
mentioned tools are available and accessible then start the stack.

> To get all available commands of the `dev` script run `./dev --help`.

### Load sample data

The sample data spans messages and a development user.

To import the sample data run

```
./dev m reset
```

This command can also be run to have a clean starting point again.

The development user is identified by: `dev@dealog.de` // `password1234`

> Note: The application if running should be stopped. This command will drop
> all databases (event store and projection / Ecto database).

### Testing

To run the tests please use `./dev test` or `./dev test.watch`.

> Functional tests are ran in the `en` locale whereas the default locale for the
> app is `de`.

### Translations

The DEalog Backoffice strings are all in english language. These are translated
to German. To quickly get the `po` files updated run `./dev translations.update`.
After that you can easily fill in the translations. If not sure about the
wording let it be resolved during the pull request process.

### Replaying projections

The DEalog Backoffice uses the [Commanded](https://commanded.io) library
providing [CQRS](https://www.martinfowler.com/bliki/CQRS.html) support. To
replay projections locally you can run the following module on the IEx shell:

For messages:

```
DEalog IEx>>> DealogBackoffice.Support.ProjectionRebuilder.rebuild_messages()
```

For accounts:

```
DEalog IEx>>> DealogBackoffice.Support.ProjectionRebuilder.rebuild_accounts()
```

These commands purge all **local** projections and replay the events.

## Collaboration

If you want to collaborate please follow this guide:

- Fork the repository
- Implement the feature, change or bugfix
- Add tests where it makes sense
- Update the [CHANGELOG.md](CHANGELOG.md)
- Update the translations (see translation section)
- Create a pull request against the original repository

## Building a release

The DEalog Backoffice can be deployed as a Docker container. The following
steps are needed to have a deployable version.

### Building the image

To build the image run:

```
docker build -t dealog/backoffice:latest .
```

This will prepare and build the image that then can be used to spawn a
running container.

The following environment variables need to be set (and respective service need
to be reachable):

- `DATABASE_URL`: The content / projection database. (f.e.: `ecto://postgres:postgres@db/bo_db_prod`)
- `EVENT_STORE_DATABASE_URL`: The database for the event store. (e.g. `ecto://postgres:postgres@db/bo_es_db_prod`)
- `SECRET_KEY_BASE`: The Phoenix secret. (generate via `mix phx.gen.secret`)
- `HOSTNAME`: The hostname needed for WebSockets. (f.e. `localhost`)
- `PROJECTORS` (optional): The projectors that should be active. Can either be `all` (default) or `local`.
- `KAFKA_HOSTS`: The Kafka host(s). (e.g. `localhost:9092`); Can be an empty string if `local` is used.
- `SMTP_RELAY`: The SMTP relay address. (f.e. `mail.domain.tld`)
- `SMTP_PORT`: The SMTP server port. (e.g. `587`)
- `SMTP_USERNAME`: The username for the SMTP server.
- `SMTP_PASSWORD`: The password for the SMTP server.

> To test this locally it is recommended to start up a PostgreSQL instance being
> accessible from the Backoffice Docker container and use a `.env` file for the
> variables mentioned above.

The following environment variables can be set:

- `SUPER_USER_EMAIL`: The email address of the super user to create.
- `SUPER_USER_PASSWORD`: The password for the super user to be set.

> If these variables are not set everything will work fine just some
> functionality will not be executed.

### Deployment all in one

For now there is a command script under `rel/run.sh` running all the setup
and migration commands on container start to ease the initial deployment.

For this to work start the container like this:

```
docker run -it -p 5000:4000 --network my_network --env-file=./.env dealog/backoffice:latest
```

> Note: You need to pass the `-it` flag to capture a `ctrl-c` interupt. If not
> passed you'd need to kill the running container.

### Setup the event store

To initialize the event store the following command needs to be run:

```
docker run --network my_network --env-file=./.env dealog/backoffice:latest bin/backoffice eval "DealogBackoffice.Release.init_event_store()"
```

> Note: `my_network` should contain the network the PostgreSQL instance is
> reachable from.

### Running migrations

```
docker run --network my_network --env-file=./.env dealog/backoffice:latest bin/backoffice eval "DealogBackoffice.Release.migrate()"
```

> Note: `my_network` should contain the network the PostgreSQL instance is
> reachable from.

### Import the adminsitrative areas

```
docker run --network my_network --env-file=./.env dealog/backoffice:latest bin/backoffice eval "DealogBackoffice.Release.import_administrative_areas()"
```

> Note: `my_network` should contain the network the PostgreSQL instance is
> reachable from.

### Create the super user

```
docker run --network my_network --env-file=./.env dealog/backoffice:latest bin/backoffice eval "DealogBackoffice.Release.create_super_user()"
```

> Note: `my_network` should contain the network the PostgreSQL instance is
> reachable from. Further ensure that the `SUPER_USER_EMAIL` and
> `SUPER_USER_PASSWORD` env vars are set.

### Start the container

```
docker run -p 5000:4000 --network my_network --env-file=./.env dealog/backoffice:latest
```

When started on locahost the DEalog Backoffice is reachable via
`http://localhost:5000`.

## Preview deployment

In order to deploy a branch for review there is a script available. The script
creates a Dokku application and deploys the current branch.

Currently this is a manual task.

> Please copy the `preview_env.dist` to `.preview_env` and set the variables
> accordingly:
> `cp preview_env.dist .preview_env`

To deploy run: `./preview_deploy` on the respective branch.
To cleanup run: `./preview_deploy --cleanup` on the branch. Additionally you
can pass a sanitized branch name as second argument to remove a different
deployment.

> You need to have access to the Dev server instance to deploy.

## Used technologies

The DEalog Backoffice application uses the following (main) technologies,
frameworks and libraries:

- [Elixir](https://elixir-lang.org)
- [Phoenix Framework](https://phoenixframework.org)
- [Commanded](https://commanded.io)
- [TailwindCSS](https://tailwindcss.com)
- [Alpine.js](https://github.com/alpinejs/alpine)
- [Postgres](https://www.postgresql.org)

## Technical decisions

As the DEalog platform transports important information and it is important to
know what happened when by whom event sourcing is implemented which serves as
an audit log as well as a reproducible state of projected data.

To get a snappy UI feeling the project leverages
[Phoenix LiveView](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html).

Mailing is done via the great [Swoosh](https://hexdocs.pm/swoosh/Swoosh.html)
library.


CHANGELOG.md
## Unreleased

- `NEW` Add administrative area filter to approvals
- `NEW` Add administrative filter to my organization
- `FIX` Fix form field styles
- `NEW` ARS IDs are projected for filtering
- `NEW` Add export of ARS for Heilbronn
- `NEW` The confirmation email can be resent
- `FIX` Organization is sent correctly to the message service
- `NEW` Messages can have a category attached
- `FIX` Add missing event store migration
- `CNG` Improve preview deployments
- `CNG` Update Elixir dependencies
- `FIX` Fix publishing ARS and organization via Kafka

## 0.5.0

This release provides the user system.

- `FIX` Sending message for approval is possible again
- `NEW` All message actions are disabled when the account is incomplete
- `NEW` User gets a warning when account is not setup
- `NEW` Author and organization meta data is saved for all message related events
- `NEW` Check origin config for websocket connections is introduced
- `CNG` Move hostname config to release
- `NEW` Release startup actions are logged properly
- `FIX` The updated date on the message for approval projection was fixed
- `FIX` The updated date on the created message projection was fixed
- `NEW` Support module to rebuild account projections was added
- `NEW` Support module to rebuild message projections was added
- `REM` Rebuild projections script was removed
- `NEW` Sample data can be loaded to the application for development
- `NEW` Changed published messages can be reverted to published state and archived
- `NEW` Changes on published messages can be discarded
- `NEW` Published messages can be archived (taken offline)
- `CNG` Changelog is cleaned up
- `CNG` Administrative areas are preloaded differently
- `CNG` Login and other user related pages are centered on the screen
- `NEW` The category of the message being published is added (hard coded at the moment)
- `NEW` The organization of the message being published is added (hard coded at the moment)
- `NEW` The ars code of the message being published is added (hard coded at the moment)
- `NEW` A super user can be created if email and password is supplied
- `NEW` Publishing of messsages to Kafka is logged
- `NEW` All command dispatches are logged
- `FIX` Publish at is sent as a timestamp instead of a ISO 8601 string
- `FIX` Type of first time published messages is changed to `Created`
- `NEW` Messages that already have been published are detected as updated ones
- `FIX` Message payload is now compatible with the Message Service
- `FIX` Project information (readme) can be viewed on released instances
- `NEW` Users are anonymously tracked to get some metrics
- `NEW` User accounts can be changed
- `NEW` Users can be onboarded with an account containing the personal data as well as the organizational settings
- `NEW` Users are listed in the settings area
- `CNG` Frontend dependecies are updated
- `FIX` Login with unconfirmed users issues is handled by emitting a user or password wrong message
- `FIX` The Kafka configuration is read correctly when no projectors are configured
- `NEW` Setup GH Actions to deploy application to AWS
- `CNG` Send ARS instead of a geocode to Kafka
- `CNG` Structure of administrative areas is changed to ARS (from AGS)
- `NEW` Mail support for registration, password reset, email change is available
- `CNG` Update dependencies
- `NEW` Authentication is implemented for all content pages
- `NEW` Users can change their email address
- `NEW` Users can change their password
- `NEW` Users can authenticate
- `NEW` Users can register

## 0.4.0

This release covers message publishing for a single organization.

- `NEW` Forms have cancel buttons
- `CNG` Deployments are update to use the new default branch `main`
- `CNG` Adapt dev script for new default branch `main`
- `CNG` Import of administrative areas is running on container start (until there is support for one-off tasks)
- `NEW` Administrative areas can be imported
- `FIX` The Kafka host can be configured via an env var
- `NEW` Published message details can be viewed
- `NEW` Published messages are shown in the all messages list
- `NEW` Preview deployment cleanup supports custom branches
- `NEW` Published messages get distributed to Kafka if enabled
- `NEW` Messages that are approved can be published
- `CNG` Projection reset script can be run in prod as well (with a security flag)
- `FIX` Tag name for master build
- `FIX` Fix GH Action output for deployment
- `NEW` Manual preview deployments are available
- `NEW` A health endpoint is added for monitoring
- `NEW` Application can be deployed via mix release on Docker
- `CNG` Rejected messages are removed from the approval section
- `CNG` Improve message listings
- `NEW` Messages can be deleted

## 0.3.0

This release has a basic approval process for a single organization on board.

- `CNG` Redirects of actions on message refer to the detail view of that message
- `NEW` Approved messages can have a note attached to them
- `NEW` Messages waiting for approval can be approved
- `NEW` Message details can be viewed for approvals
- `NEW` Message details can be viewed for my organization
- `NEW` Rejected messages can have a reason
- `NEW` Messages waiting for approval can be rejected
- `NEW` Dev script has a task to reset the task database
- `NEW` Messages waiting for approval are listed in the approvals section
- `NEW` Messages can be sent for approval
- `NEW` Error messages are displayed as notification
- `CNG` No change saves do not emit a changed event
- `NEW` Favicons are added
- `CNG` Improve my organization table layout

## 0.2.0

This release contains adding and changing new messages for a single organization.

- `NEW` Database is setup and provisioned on deployment
- `NEW` Projections can be reset more easily
- `CNG` Mix tasks can be ran independently of "stack" container
- `NEW` Messages can be changed
- `NEW` Dev script supports spawning an IEx session
- `NEW` Dev script supports running arbitrary mix commands

config/releases.exs
import Config

database_url =
  System.get_env("DATABASE_URL") ||
    raise """
    environment variable DATABASE_URL is missing.
    For example: ecto://USER:PASS@HOST/DATABASE
    """

event_store_database_url =
  System.get_env("EVENT_STORE_DATABASE_URL") ||
    raise """
    environment variable EVENT_STORE_DATABASE_URL is missing.
    For example: ecto://USER:PASS@HOST/DATABASE
    """

config :dealog_backoffice, DealogBackoffice.Repo,
  migration_timestamps: [type: :utc_datetime_usec],
  # ssl: true,
  url: database_url,
  pool_size: String.to_integer(System.get_env("POOL_SIZE") || "10")

config :dealog_backoffice, DealogBackoffice.EventStore,
  # ssl: true,
  url: event_store_database_url,
  pool_size: String.to_integer(System.get_env("POOL_SIZE") || "10")

config :dealog_backoffice, :projection,
  projectors: String.to_existing_atom(System.get_env("PROJECTORS") || "all")

kafka_hosts =
  System.get_env("KAFKA_HOSTS") ||
    raise """
        environment variable KAFKA_HOSTS is missing.
        For example: localhost:9092,localhost:9093,localhost:9094
    """

config :kafka_ex,
  brokers: kafka_hosts

secret_key_base =
  System.get_env("SECRET_KEY_BASE") ||
    raise """
    environment variable SECRET_KEY_BASE is missing.
    You can generate one by calling: mix phx.gen.secret
    """

hostname =
  System.get_env("HOSTNAME") ||
    raise """
    environment variable HOSTNAME is missing.
    """

config :dealog_backoffice, DealogBackofficeWeb.Endpoint,
  url: [scheme: "https", host: hostname, port: 443],
  http: [
    port: String.to_integer(System.get_env("PORT") || "4000"),
    transport_options: [socket_opts: [:inet6]]
  ],
  check_origin: ["//#{hostname}"],
  secret_key_base: secret_key_base

smtp_relay =
  System.get_env("SMTP_RELAY") ||
    raise """
        environment variable SMTP_RELAY is missing.
        For example: mail.domain.tld
    """

smtp_port =
  System.get_env("SMTP_PORT") ||
    raise """
        environment variable SMTP_PORT is missing.
        For example: 587
    """

smtp_username =
  System.get_env("SMTP_USERNAME") ||
    raise """
        environment variable SMTP_USERNAME is missing.
        For example: mail@domain.tld
    """

smtp_password =
  System.get_env("SMTP_PASSWORD") ||
    raise """
        environment variable SMTP_PASSWORD is missing.
        For example: s3c3tP4ssw0Rd
    """

# Swoosh mailing
config :dealog_backoffice, DealogBackoffice.Accounts.Mailer,
  adapter: Swoosh.Adapters.SMTP,
  relay: smtp_relay,
  port: smtp_port,
  username: smtp_username,
  password: smtp_password

# Super user settings
config :dealog_backoffice, :super_user,
  email: System.get_env("SUPER_USER_EMAIL"),
  password: System.get_env("SUPER_USER_PASSWORD")
